rust/src/kernel/actions/arrow.rs

```diff
@@ -0,0 +1,275 @@
+                    ArrowField::new(
+                        "value",
+                        ArrowDataType::try_from(a.value_type())?,
+                        a.value_contains_null(),
+                    ),
+                ]
+                .into(),
+            ),
+            false, // always non-null
+        ))
+    }
+}
+
+impl TryFrom<&DataType> for ArrowDataType {
+    type Error = ArrowError;
+
+    fn try_from(t: &DataType) -> Result<Self, ArrowError> {
+        match t {
+            DataType::Primitive(p) => {
+                match p {
+                    PrimitiveType::String => Ok(ArrowDataType::Utf8),
+                    PrimitiveType::Long => Ok(ArrowDataType::Int64), // undocumented type
+                    PrimitiveType::Integer => Ok(ArrowDataType::Int32),
+                    PrimitiveType::Short => Ok(ArrowDataType::Int16),
+                    PrimitiveType::Byte => Ok(ArrowDataType::Int8),
+                    PrimitiveType::Float => Ok(ArrowDataType::Float32),
+                    PrimitiveType::Double => Ok(ArrowDataType::Float64),
+                    PrimitiveType::Boolean => Ok(ArrowDataType::Boolean),
+                    PrimitiveType::Binary => Ok(ArrowDataType::Binary),
+                    PrimitiveType::Decimal(precision, scale) => {
+                        let precision = u8::try_from(*precision).map_err(|_| {
+                            ArrowError::SchemaError(format!(
+                                "Invalid precision for decimal: {}",
+                                precision
+                            ))
+                        })?;
+                        let scale = i8::try_from(*scale).map_err(|_| {
+                            ArrowError::SchemaError(format!("Invalid scale for decimal: {}", scale))
+                        })?;
+
+                        if precision <= 38 {
+                            Ok(ArrowDataType::Decimal128(precision, scale))
+                        } else if precision <= 76 {
+                            Ok(ArrowDataType::Decimal256(precision, scale))
+                        } else {
+                            Err(ArrowError::SchemaError(format!(
+                                "Precision too large to be represented in Arrow: {}",
+                                precision
+                            )))
+                        }
+                    }
+                    PrimitiveType::Date => {
+                        // A calendar date, represented as a year-month-day triple without a
+                        // timezone. Stored as 4 bytes integer representing days since 1970-01-01
+                        Ok(ArrowDataType::Date32)
+                    }
+                    PrimitiveType::Timestamp => {
+                        // Issue: https://github.com/delta-io/delta/issues/643
+                        Ok(ArrowDataType::Timestamp(TimeUnit::Microsecond, None))
```

Comment:
What about timestamps that were written with the UTC timezone? Are we doing anything with the IsAdjustedToUtc flag in the parquet timestamp column?

Reply:
right now this is not handled. Within delta this is also gated behind the `timestampNtz` feature. As this PR is quite hard to review due to its size, I focussed on not adjusting any business logic but rather only update the schema / action definitions and getting things to build. So support for that feature should be in a follow-up PR.